csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,233 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/move_pad.h>
+
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+#include <transform_replay.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+struct Edge {
+  Expr* expr_ = nullptr;
+  size_t index_ = 0;
+
+  Edge(Expr* expr, size_t index) : expr_(expr), index_(index) {}
+
+  Val* val() const {
+    return expr_->input(index_);
+  }
+};
+
+Val* propagatePadToProducer(PadOp* pad_op) {
+  std::vector<Val*> pad_dependencies;
+
+  auto candidate_check = [&pad_dependencies](Val* val) {
+    if (!val->isA<TensorView>()) {
+      return false;
+    }
+    if (val->uses().size() > 1) {
+      return false;
+    }
+    if (val->isFusionOutput()) {
+      return false;
+    }
+    if (std::any_of(
+            pad_dependencies.begin(),
+            pad_dependencies.end(),
+            [val](Val* pad_dependency) {
+              return DependencyCheck::isDependencyOf(pad_dependency, val);
+            })) {
+      return false;
+    }
+    return true;
+  };
+
+  // NOTE: the optimization logic assumes a zero pad_op.
+  // This is used for logic in handling binary operations, we should extend this
+  // later.
+  if (!pad_op->value()->isZero()) {
+    return nullptr;
+  }
+
+  if (!candidate_check(pad_op->in())) {
+    return nullptr;
+  }
+
+  for (Val* val : pad_op->inputs()) {
+    if (val == pad_op->in() || val->isConst()) {
+      continue;
+    }
+    pad_dependencies.push_back(val);
+  }
+
+  std::vector<Edge> frontier;
+  // TODO: not sure if I need a stack if I need to keep a replay_sequence.
+  std::stack<Edge> stack;
+  std::vector<Expr*> replay_sequence;
+  stack.emplace(pad_op, 0);
+
+  // tvs in stack are:
+  //   1. single use;
+  //   2. not an output;
+  //   3. cleared dependency of `pad_depdencies;
+  //   4. maybe also check aliases?!
+  while (!stack.empty()) {
+    Edge edge = stack.top();
+    Expr* def = edge.val()->definition();
+    stack.pop();
+
+    if (def->isA<UnaryOp>()) {
+      auto* uop = def->as<UnaryOp>();
+      // TODO: exception to break propagation.
+      if (candidate_check(uop->in())) {
```

Comment:
need to exclude some operation types here to avoid illegal instructions like division by 0.